src/packageurl/contrib/url2purl.py

```diff
@@ -276,3 +276,22 @@ def build_sourceforge_purl(uri):
 @purl_router.route(cargo_pattern)
 def build_cargo_purl(uri):
     return purl_from_pattern('cargo', cargo_pattern, uri)
+
+
+@purl_router.route('https?://github.com.*/.*')
+def build_github_purl(uri):
+    path = unquote_plus(urlparse(uri).path)
+    segments = [seg for seg in path.split('/') if seg]
+    
+    if len(segments) == 2:
+        namespace = segments[0]
+        name = segments[1]
+        return PackageURL('github', name=name, namespace=namespace)
+    
+    if len(segments) >= 4:
+        if segments[2] == 'tree':
+            namespace = segments[0]
+            name = segments[1]
+            version = segments[3]
+            subpath = '/'.join(segments[4:])
+            return PackageURL('github', name=name, namespace=namespace, version=version, subpath=subpath)
```

Comment:
What happens if `len(segments) == 1` or `len(segments) == 3`?

Reply:
I have covered a test case for it when it will be 1, it will be null, and for a valid GitHub URL, if it's having more than three elements the third element should be tree, and the fourth need to be branch.

Reply:
If it will be having only one in segment then it's only the namespace, and namespace won't be pointing towards a repo, I will be helpful if you can provide me with some test cases that I should include

Reply:
Hi @majurg, I have added test cases where segment length is 3, so how does it look now :D, Please suggest me if any other changes are required.

Reply:
The test cases cover it fine and that is the expected behavior. However, I would want that reflected in the code for clarity:
```
if len(segments) == 1 or len(segments) == 3:
    # These are not valid GitHub URLs to a specific package, so return null
    return None
```

Reply:
Sure, got you mate :grinning: , making the following changes

Reply:
Done @majurg , thanks for your guidance :D

Reply:
Are there more changes to be done ? If it's looking good I can start working for BitBucket URLs in the same manner

Reply:
@TG1999 yes you can add bitbucket to this same PR, just so I dont have to search around too much xD

Reply:
Cool !! Should I make a new commit for it in this PR or squash the BitBucket work in this commit ?

Reply:
I think have 2 seperate commits is the way to go:
1): GitHub commit
2): BitBucket commit

And then just rebase of your BitBucket commit any changes for either.

Reply:
Sure I will be sending my first draft in sometime

Reply:
@majurg , I have added commit for bitbucket URLs it was almost similar to th GitHub one, so it did not take too much time. Please review it so we can then move on the last one i.e GitLab

Reply:
In this case, if len(segments) == 3 won't return None everytime, if segments[2] == 'src' then the given URL will be valid, for example https://bitbucket.org/TG1999/first_repo/src/

Reply:
@TG1999 We should have separate functions for both github and bitbucket, since they are a little different (as you state above)

Reply:
Okay, got this one

Reply:
@majurg done, please review this :D